<a href="https://colab.research.google.com/github/shailesh3335/Telecom-Churn-Analysis/blob/main/EDA_Telecom_Churn_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Project Name**- Telecom Churn Analysis


##### **Project Type**    - EDA
##### **Contribution**    - Individual
##### **Name - Shailesh Kakade**


# **Project Summary -**

# **GitHub Link -**https://github.com/shailesh3335/Telecom-Churn-Analysis

In [47]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [52]:
df = pd.read_csv("/content/Telecom Churn.csv")
df.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

In [54]:
df.shape

(3333, 20)

In [55]:
df.columns

Index(['State', 'Account length', 'Area code', 'International plan',
       'Voice mail plan', 'Number vmail messages', 'Total day minutes',
       'Total day calls', 'Total day charge', 'Total eve minutes',
       'Total eve calls', 'Total eve charge', 'Total night minutes',
       'Total night calls', 'Total night charge', 'Total intl minutes',
       'Total intl calls', 'Total intl charge', 'Customer service calls',
       'Churn'],
      dtype='object')

### Variables Description 

* **State                :**categorica for the 50 states

* **Account Length       :**how long account has been active

* **Area Code            :**Code Number of Area having some States included in each area code

* **lntl Plan            :**Internat ional plan activated ( yes, no )

* **VMail Plan           :**  ice Mail plan activated ( yes ,no )

* **VMail Message        :**No.of voice mail messages

* **Day Mins             :**Total day minutes used

* **Day calls**         :Total day calls made

* **Day Charge**         :Total day charge

* **Eve Mins**          :Total evening minutes

* **Eve Calls**          :Total evening calls

* **Eve Charge**         :Total evening charge

* **Night Mins**         :Total night minutes

* **Night Calls**        :Total night calls

* **Night Charge**      :Total night charge

* **Intl Mins**         :Total International minutes used

* **Intl Calls**         :Total International calls made

* **Intl Charge**        :Total International charge

* **CustServ calls**    :Number of customer service caUs made

* **Churn**             :Customer churn (Target Variable True=1, False=0)

In [56]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3333 non-null   object 
 4   Voice mail plan         3333 non-null   object 
 5   Number vmail messages   3333 non-null   int64  
 6   Total day minutes       3333 non-null   float64
 7   Total day calls         3333 non-null   int64  
 8   Total day charge        3333 non-null   float64
 9   Total eve minutes       3333 non-null   float64
 10  Total eve calls         3333 non-null   int64  
 11  Total eve charge        3333 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3333 non-null   int64  
 14  Total night charge      3333 non-null   

In [57]:
df.isnull().sum()

State                     0
Account length            0
Area code                 0
International plan        0
Voice mail plan           0
Number vmail messages     0
Total day minutes         0
Total day calls           0
Total day charge          0
Total eve minutes         0
Total eve calls           0
Total eve charge          0
Total night minutes       0
Total night calls         0
Total night charge        0
Total intl minutes        0
Total intl calls          0
Total intl charge         0
Customer service calls    0
Churn                     0
dtype: int64

 it can be concluded that there are no missing values (null values) in any of the columns of the DataFrame df. Each column has a count of 0, indicating that there are no missing values in any of the columns.

Therefore, the dataset appears to be complete, with no missing data in any of the columns.



In [58]:
df["Churn"] = df["Churn"].astype("int64")

In [59]:
df.describe()

Account length    Area code  Number vmail messages  Total day minutes  \
count     3333.000000  3333.000000            3333.000000        3333.000000   
mean       101.064806   437.182418               8.099010         179.775098   
std         39.822106    42.371290              13.688365          54.467389   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.400000   
max        243.000000   510.000000              51.000000         350.800000   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
count      3333.000000       3333.000000        3333.000000      3333.000000   
mean        100.435644         30.562307         200.980348       100.114311   
std          20.069084          9.259435          50.713844        19.922625   
min           0.000000          0.000000           0.000000         0.000000   
25%          87.000000         24.430000         166.600000        87.000000   
50%         101.000000         30.500000         201.400000       100.000000   
75%         114.000000         36.790000         235.300000       114.000000   
max         165.000000         59.640000         363.700000       170.000000   

       Total eve charge  Total night minutes  Total night calls  \
count       3333.000000          3333.000000        3333.000000   
mean          17.083540           200.872037         100.107711   
std            4.310668            50.573847          19.568609   
min            0.000000            23.200000          33.000000   
25%           14.160000           167.000000          87.000000   
50%           17.120000           201.200000         100.000000   
75%           20.000000           235.300000         113.000000   
max           30.910000           395.000000         175.000000   

       Total night charge  Total intl minutes  Total intl calls  \
count         3333.000000         3333.000000       3333.000000   
mean             9.039325           10.237294          4.479448   
std              2.275873            2.791840          2.461214   
min              1.040000            0.000000          0.000000   
25%              7.520000            8.500000          3.000000   
50%              9.050000           10.300000          4.000000   
75%             10.590000           12.100000          6.000000   
max             17.770000           20.000000         20.000000   

       Total intl charge  Customer service calls        Churn  
count        3333.000000             3333.000000  3333.000000  
mean            2.764581                1.562856     0.144914  
std             0.753773                1.315491     0.352067  
min             0.000000                0.000000     0.000000  
25%             2.300000                1.000000     0.000000  
50%             2.780000                1.000000     0.000000  
75%             3.270000                2.000000     0.000000  
max             5.400000                9.000000     1.000000

   1) Count: The count for each column is 3333, indicating that there are 3333 observations in the dataset.

   2) Mean: The mean provides the average value for each column. For example, the average account length is approximately 101.06, the average number of voicemail messages is approximately 8.10, and so on.

   3) Standard Deviation (std): The standard deviation measures the variability or dispersion of data around the mean. A higher standard deviation indicates greater variability in the data.

  4) Minimum (min): The minimum value represents the smallest value observed in each column. For instance, the minimum account length is 1, the minimum number of voicemail messages is 0, and so on.

  5) 25th Percentile (25%): The 25th percentile is the value below which 25% of the data falls. It provides information about the lower quartile of the data distribution.

   6) 50th Percentile (50%)/Median: The 50th percentile, also known as the median, divides the data into two equal halves. It represents the middle value of the dataset.

   7) 75th Percentile (75%): The 75th percentile is the value below which 75% of the data falls. It provides information about the upper quartile of the data distribution.

   8) Maximum (max): The maximum value represents the largest value observed in each column. For example, the maximum account length is 243, the maximum number of voicemail messages is 51, and so on.

These descriptive statistics provide insights into the central tendency, variability, and range of the dataset across different columns.




In [9]:
df.describe(include=["object", "bool"])


State International plan Voice mail plan
count   3333               3333            3333
unique    51                  2               2
top       WV                 No              No
freq     106               3010            2411

conclusions can be made for the categorical columns of the DataFrame:

    State:
        The count is 3333, indicating that there are 3333 observations in this column.
        There are 51 unique states represented in the dataset.
        The top state is WV (West Virginia) with a frequency of 106 occurrences.

    International plan:
        The count is 3333, indicating that there are 3333 observations in this column.
        There are 2 unique categories for international plan: Yes and No.
        The top category is No with a frequency of 3010 occurrences.

    Voice mail plan:
        The count is 3333, indicating that there are 3333 observations in this column.
        There are 2 unique categories for voice mail plan: Yes and No.
        The top category is No with a frequency of 2411 occurrences.

These conclusions provide information about the distribution and frequencies of the categorical variables in the DataFrame. It shows the number of unique categories, the most frequent categories, and their corresponding frequencies.


In [10]:
df["Churn"].value_counts()

0    2850
1     483
Name: Churn, dtype: int64

In [36]:
df["Churn"].mean()

0.14491449144914492

14.49% is actually quite bad for a company; such a churn rate can make the company go bankrupt.


Let's use it to answer the question:

What are average values of numerical features for churned users?

In [11]:
#The average values of different features for churned customers, allowing for a better understanding of their behavior and characteristics.
df[df["Churn"] == 1].mean()

<ipython-input-11-c961b1669f63>:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df[df["Churn"] == 1].mean()


Account length            102.664596
Area code                 437.817805
Number vmail messages       5.115942
Total day minutes         206.914079
Total day calls           101.335404
Total day charge           35.175921
Total eve minutes         212.410145
Total eve calls           100.561077
Total eve charge           18.054969
Total night minutes       205.231677
Total night calls         100.399586
Total night charge          9.235528
Total intl minutes         10.700000
Total intl calls            4.163561
Total intl charge           2.889545
Customer service calls      2.229814
Churn                       1.000000
dtype: float64

In [13]:
df.sort_values(by=["Churn", "Total day charge"], ascending=[True, False]).head()

State  Account length  Area code International plan Voice mail plan  \
688     MN              13        510                 No             Yes   
2259    NC             210        415                 No             Yes   
534     LA              67        510                 No              No   
575     SD             114        415                 No             Yes   
2858    AL             141        510                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
688                      21              315.6              105   
2259                     31              313.8               87   
534                       0              310.4               97   
575                      36              309.9               90   
2858                     28              308.0              123   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
688              53.65              208.9               71             17.76   
2259             53.35              147.7              103             12.55   
534              52.77               66.5              123              5.65   
575              52.68              200.3               89             17.03   
2858             52.36              247.8              128             21.06   

      Total night minutes  Total night calls  Total night charge  \
688                 260.1                123               11.70   
2259                192.7                 97                8.67   
534                 246.5                 99               11.09   
575                 183.5                105                8.26   
2858                152.9                103                6.88   

      Total intl minutes  Total intl calls  Total intl charge  \
688                 12.1                 3               3.27   
2259                10.1                 7               2.73   
534                  9.2                10               2.48   
575                 14.2                 2               3.83   
2858                 7.4                 3               2.00   

      Customer service calls  Churn  
688                        3      0  
2259                       3      0  
534                        4      0  
575                        1      0  
2858                       1      0

In [15]:
# Churn data groupby Area Code Wise (The churn counts categorized by area code. It shows the number of churned and non-churned customers for each area code.)
pd.DataFrame(df.groupby('Area code')['Churn'].value_counts().reset_index(name="Count"))

Area code  Churn  Count
0        408      0    716
1        408      1    122
2        415      0   1419
3        415      1    236
4        510      0    715
5        510      1    125

In [17]:
# Churn Counts grouby State wise
pd.DataFrame(df.groupby('State')['Churn'].value_counts(  ).reset_index(name="Count"))

State  Churn  Count
0      AK      0     49
1      AK      1      3
2      AL      0     72
3      AL      1      8
4      AR      0     44
..    ...    ...    ...
97     WI      1      7
98     WV      0     96
99     WV      1     10
100    WY      0     68
101    WY      1      9

[102 rows x 3 columns]

In [28]:
columns_to_show = ["Total day minutes", "Total eve minutes", "Total night minutes"]

df.groupby(["Churn"])[columns_to_show].describe(percentiles=[])

Total day minutes                                            \
                  count        mean        std  min    50%    max   
Churn                                                               
0                2850.0  175.175754  50.181655  0.0  177.2  315.6   
1                 483.0  206.914079  68.997792  0.0  217.6  350.8   

      Total eve minutes                                             \
                  count        mean        std   min    50%    max   
Churn                                                                
0                2850.0  199.043298  50.292175   0.0  199.6  361.8   
1                 483.0  212.410145  51.728910  70.9  211.3  363.7   

      Total night minutes                                              
                    count        mean        std   min     50%    max  
Churn                                                                  
0                  2850.0  200.133193  51.105032  23.2  200.25  395.0  
1                   483.0  205.231677  47.132825  47.4  204.80  354.9

These conclusions provide insights into the descriptive statistics of the total day, evening, and night minutes for churned and non-churned customers. They highlight the differences in means, standard deviations, minimum and maximum values, and medians between the two groups, allowing for a comparison of their phone usage patterns.

In [29]:
columns_to_show = ["Total day minutes", "Total eve minutes", "Total night minutes"]

df.groupby(["Churn"])[columns_to_show].agg([np.mean, np.std, np.min, np.max])

Total day minutes                        Total eve minutes             \
                   mean        std amin   amax              mean        std   
Churn                                                                         
0            175.175754  50.181655  0.0  315.6        199.043298  50.292175   
1            206.914079  68.997792  0.0  350.8        212.410145  51.728910   

                   Total night minutes                          
       amin   amax                mean        std  amin   amax  
Churn                                                           
0       0.0  361.8          200.133193  51.105032  23.2  395.0  
1      70.9  363.7          205.231677  47.132825  47.4  354.9

In [30]:
pd.crosstab(df["Churn"], df["International plan"])


International plan    No  Yes
Churn                        
0                   2664  186
1                    346  137

These conclusions provide insights into the relationship between the churn status and the presence of an international plan. It indicates that a higher proportion of churned customers had an international plan compared to non-churned customers. This suggests that the presence of an international plan may have an influence on customer churn.

In [31]:
pd.crosstab(df["Churn"], df["Voice mail plan"], normalize=True)

Voice mail plan        No       Yes
Churn                              
0                0.602460  0.252625
1                0.120912  0.024002

These conclusions provide insights into the relationship between the churn status and the presence of a voice mail plan. It indicates that a higher proportion of customers who did not churn did not have a voice mail plan, while a smaller proportion had a voice mail plan. In contrast, a lower proportion of churned customers did not have a voice mail plan, while a very small proportion had a voice mail plan. 

In [32]:
df.pivot_table(
    ["Total day calls", "Total eve calls", "Total night calls"],
    ["Area code"],
    aggfunc="mean",
)

Total day calls  Total eve calls  Total night calls
Area code                                                     
408             100.496420        99.788783          99.039379
415             100.576435       100.503927         100.398187
510             100.097619        99.671429         100.601190

These conclusions provide insights into the average number of calls made during the day, evening, and night for each area code. It shows that the mean values of call counts are relatively similar across the different area codes, indicating a balanced distribution of call activity throughout the day.

Let's see how churn rate is related to the International plan feature.

In [35]:
pd.crosstab(df["Churn"], df["International plan"], margins=True)

International plan    No  Yes   All
Churn                              
0                   2664  186  2850
1                    346  137   483
All                 3010  323  3333

In [63]:
# Assigning churn true customers data to variable df_churn
df_churn=df[(df['Churn']==1)]
df_churn.head()

State  Account length  Area code International plan Voice mail plan  \
10    IN              65        415                 No              No   
15    NY             161        415                 No              No   
21    CO              77        408                 No              No   
33    AZ              12        408                 No              No   
41    MD             135        408                Yes             Yes   

    Number vmail messages  Total day minutes  Total day calls  \
10                      0              129.1              137   
15                      0              332.9               67   
21                      0               62.4               89   
33                      0              249.6              118   
41                     41              173.1               85   

    Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
10             21.95              228.5               83             19.42   
15             56.59              317.8               97             27.01   
21             10.61              169.9              121             14.44   
33             42.43              252.4              119             21.45   
41             29.43              203.9              107             17.33   

    Total night minutes  Total night calls  Total night charge  \
10                208.8                111                9.40   
15                160.6                128                7.23   
21                209.6                 64                9.43   
33                280.2                 90               12.61   
41                122.2                 78                5.50   

    Total intl minutes  Total intl calls  Total intl charge  \
10                12.7                 6               3.43   
15                 5.4                 9               1.46   
21                 5.7                 6               1.54   
33                11.8                 3               3.19   
41                14.6                15               3.94   

    Customer service calls  Churn  
10                       4      1  
15                       4      1  
21                       5      1  
33                       1      1  
41                       0      1

In [62]:
df_churn.shape

(483, 20)

In [64]:
# Assigning Customers data with no international plan and having voice mail plan
df_true_voice=df_churn[(df_churn['International plan']=='No')&(df_churn['Voice mail plan']=='Yes')]
df_true_voice.describe()

Account length   Area code  Number vmail messages  Total day minutes  \
count       44.000000   44.000000              44.000000          44.000000   
mean       100.863636  441.000000              31.818182         163.306818   
std         35.580626   45.294386               6.728187          47.918363   
min         17.000000  408.000000              17.000000          82.300000   
25%         77.000000  408.000000              28.000000         133.275000   
50%        103.000000  415.000000              32.000000         167.350000   
75%        123.000000  510.000000              35.250000         191.675000   
max        174.000000  510.000000              48.000000         322.400000   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
count        44.000000         44.000000          44.000000        44.000000   
mean        100.977273         27.762727         195.122727       101.363636   
std          20.274269          8.146506          52.169344        20.291012   
min          59.000000         13.990000          75.300000        56.000000   
25%          86.500000         22.655000         162.250000        90.750000   
50%         100.500000         28.450000         200.250000       102.500000   
75%         114.250000         32.587500         232.525000       116.250000   
max         147.000000         54.810000         315.400000       136.000000   

       Total eve charge  Total night minutes  Total night calls  \
count         44.000000            44.000000          44.000000   
mean          16.585909           197.986364         102.704545   
std            4.434638            50.631402          19.078272   
min            6.400000            87.400000          49.000000   
25%           13.790000           166.450000          90.000000   
50%           17.025000           191.350000         103.000000   
75%           19.767500           224.400000         115.250000   
max           26.810000           321.200000         145.000000   

       Total night charge  Total intl minutes  Total intl calls  \
count           44.000000           44.000000         44.000000   
mean             8.909091           10.202273          4.363636   
std              2.277850            2.615116          1.942079   
min              3.930000            5.800000          1.000000   
25%              7.490000            8.675000          3.000000   
50%              8.615000           10.050000          4.000000   
75%             10.095000           11.800000          5.250000   
max             14.450000           15.700000         11.000000   

       Total intl charge  Customer service calls  Churn  
count          44.000000               44.000000   44.0  
mean            2.755455                3.295455    1.0  
std             0.706022                1.719648    0.0  
min             1.570000                0.000000    1.0  
25%             2.342500                2.000000    1.0  
50%             2.715000                4.000000    1.0  
75%             3.190000                4.000000    1.0  
max             4.240000                6.000000    1.0

In [65]:
# Assigning Customers data with no international plan and no voice mail plan
df_true_no=df_churn[(df_churn['International plan']=='No')&(df_churn['Voice mail plan']=='No')]
df_true_no.describe()

Account length   Area code  Number vmail messages  Total day minutes  \
count      302.000000  302.000000                  302.0         302.000000   
mean       101.910596  434.920530                    0.0         219.295695   
std         41.281249   41.027343                    0.0          71.230008   
min          1.000000  408.000000                    0.0           0.000000   
25%         75.250000  408.000000                    0.0         158.025000   
50%        100.500000  415.000000                    0.0         237.950000   
75%        126.750000  415.000000                    0.0         272.700000   
max        225.000000  510.000000                    0.0         350.800000   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
count       302.000000        302.000000         302.000000       302.000000   
mean        102.546358         37.280795         216.337086       100.947020   
std          22.122346         12.109283          50.205869        20.236242   
min           0.000000          0.000000          70.900000        48.000000   
25%          89.000000         26.862500         180.625000        87.000000   
50%         104.000000         40.450000         216.250000       102.000000   
75%         118.000000         46.360000         251.250000       114.000000   
max         165.000000         59.640000         350.500000       168.000000   

       Total eve charge  Total night minutes  Total night calls  \
count        302.000000           302.000000         302.000000   
mean          18.388609           211.092053          99.811258   
std            4.267232            44.703920          20.134287   
min            6.030000            47.400000          51.000000   
25%           15.352500           178.200000          84.250000   
50%           18.385000           210.100000         100.000000   
75%           21.355000           242.700000         115.000000   
max           29.790000           354.900000         158.000000   

       Total night charge  Total intl minutes  Total intl calls  \
count          302.000000          302.000000        302.000000   
mean             9.499338           10.281457          4.235099   
std              2.011642            2.639966          2.432004   
min              2.130000            2.000000          1.000000   
25%              8.020000            8.700000          3.000000   
50%              9.450000           10.300000          4.000000   
75%             10.920000           12.075000          5.000000   
max             15.970000           18.300000         15.000000   

       Total intl charge  Customer service calls  Churn  
count         302.000000              302.000000  302.0  
mean            2.776722                2.357616    1.0  
std             0.712955                1.893666    0.0  
min             0.540000                0.000000    1.0  
25%             2.350000                1.000000    1.0  
50%             2.780000                2.000000    1.0  
75%             3.262500                4.000000    1.0  
max             4.940000                9.000000    1.0

In [66]:
# Assigning Customers data with both international plan and voice mail plan
df_true_yes=df_churn[(df_churn['International plan']=='Yes')&(df_churn['Voice mail plan']=='Yes')]
df_true_yes.describe()

Account length   Area code  Number vmail messages  Total day minutes  \
count       36.000000   36.000000              36.000000          36.000000   
mean       109.416667  442.277778              29.750000         192.813889   
std         32.514942   45.647476               7.721491          50.202886   
min         36.000000  408.000000              15.000000         103.200000   
25%         89.750000  413.250000              26.000000         156.400000   
50%        113.000000  415.000000              29.000000         196.150000   
75%        131.500000  510.000000              35.000000         222.450000   
max        188.000000  510.000000              44.000000         292.900000   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
count        36.000000         36.000000          36.000000        36.000000   
mean        101.555556         32.779722         211.811111       100.472222   
std          16.306781          8.534531          46.736794        19.395855   
min          59.000000         17.540000         107.100000        61.000000   
25%          89.750000         26.592500         181.875000        93.000000   
50%         102.000000         33.350000         209.350000        99.000000   
75%         114.250000         37.820000         253.400000       112.500000   
max         137.000000         49.790000         312.500000       133.000000   

       Total eve charge  Total night minutes  Total night calls  \
count         36.000000            36.000000          36.000000   
mean          18.004444           187.419444         101.027778   
std            3.972455            44.774456          19.549916   
min            9.100000            82.300000          67.000000   
25%           15.460000           157.750000          81.750000   
50%           17.795000           192.450000         101.500000   
75%           21.540000           214.125000         117.250000   
max           26.560000           271.700000         137.000000   

       Total night charge  Total intl minutes  Total intl calls  \
count           36.000000           36.000000         36.000000   
mean             8.433333           12.297222          4.972222   
std              2.015419            3.332565          4.003471   
min              3.700000            4.100000          2.000000   
25%              7.097500           10.600000          2.000000   
50%              8.660000           13.350000          3.000000   
75%              9.635000           14.200000          6.000000   
max             12.230000           20.000000         20.000000   

       Total intl charge  Customer service calls  Churn  
count          36.000000               36.000000   36.0  
mean            3.320278                1.722222    1.0  
std             0.899328                2.036960    0.0  
min             1.110000                0.000000    1.0  
25%             2.862500                0.000000    1.0  
50%             3.605000                1.000000    1.0  
75%             3.830000                3.000000    1.0  
max             5.400000                9.000000    1.0

In [67]:
print("Churned Customer Shape:-",df_churn.shape)
print(" ")
print("customers data with international plan and no voice plan:-",df_true_intl.shape)
print(" ")
print("Customers data with no international plan and having voice mail plan:-",df_true_voice.shape)
print(" ")
print("Customers data with no international plan and no voice mail plan:-",df_true_no.shape)
print(" ")
print("Customers data with both international plan and voice mail plan:-",df_true_yes.shape)
print(" ")

Churned Customer Shape:- (483, 20)
 
customers data with international plan and no voice plan:- (231, 20)
 
Customers data with no international plan and having voice mail plan:- (44, 20)
 
Customers data with no international plan and no voice mail plan:- (302, 20)
 
Customers data with both international plan and voice mail plan:- (36, 20)
 


In [71]:
# Assigning customers data to respective area codes
df_true_voice_415=df_true_voice[df_true_voice['Area code']==415]
df_true_voice_510=df_true_voice[df_true_voice['Area code']==510]
df_true_voice_408=df_true_voice[df_true_voice['Area code']==408]
# Assigning customers data to respective area codes
df_true_yes_415=df_true_yes[df_true_yes['Area code']==415]
df_true_yes_510=df_true_yes[df_true_yes['Area code']==510]
df_true_yes_408=df_true_yes[df_true_yes['Area code']==408]

In [72]:
# Assigning customers data with no international plan and having voice mail plan state list in area 415
df_true_voice_415_states=list(df_true_voice_415['State'].unique())
print("States list where customers data with no international plan and having voice mail plan in area 415 :-",df_true_voice_415_states)
print(" ")
# Assigning customers data  with both international plan and voice mail plan state list in area 415
df_true_yes_415_states=list(df_true_yes_415['State'].unique())
print("States List where customers data  with both international plan and voice mail plan in area 415 :-",df_true_yes_415_states)
print(" ")
# Getting Poor Network connectivity states based on our hypothetical logic
poor_netwok_states_415=set(df_true_voice_415_states).intersection(set(df_true_yes_415_states))
print("Poor Network connectivity states in area 415 :-",list(poor_netwok_states_415))

States list where customers data with no international plan and having voice mail plan in area 415 :- ['OR', 'CT', 'MS', 'LA', 'NV', 'VA', 'MT', 'CA', 'ND', 'NY', 'ID', 'NH', 'AR', 'KY', 'KS']
 
States List where customers data  with both international plan and voice mail plan in area 415 :- ['MN', 'NV', 'WI', 'KS', 'WY', 'NM', 'IL', 'DE', 'MD', 'OR', 'TX', 'ME', 'OH']
 
Poor Network connectivity states in area 415 :- ['OR', 'NV', 'KS']


In [73]:
# Assigning Customers data with no international plan and having voice mail plan state list in area 408
df_true_voice_408_states=list(df_true_voice_408['State'].unique())
print("States list Customers data with no international plan and having voice mail plan state list in area 408 :-",df_true_voice_408_states)
print(" ")
# Assigning customers data  with both international plan and voice mail plan state list in area 408
df_true_yes_408_states=list(df_true_yes_408['State'].unique())
print("States list customers data  with both international plan and voice mail plan state list in area 408 :-",df_true_yes_408_states)
print(" ")
# Getting Poor Network connectivity states based on our hypothetical logic
poor_netwok_states_408=set(df_true_voice_408_states).intersection(set(df_true_yes_408_states))
# Getting Common Poor Network connectivity states based on our hypothetical logic from both area 408 & 415
poor_network_states=poor_netwok_states_415.union(poor_netwok_states_408)
print("Poor Network connectivity states in both area 415, 408 :-",list(poor_network_states))

States list Customers data with no international plan and having voice mail plan state list in area 408 :- ['CO', 'SD', 'NM', 'GA', 'MA', 'UT', 'NC', 'WA', 'KY', 'MI', 'PA', 'NH', 'MT']
 
States list customers data  with both international plan and voice mail plan state list in area 408 :- ['MD', 'AL', 'NC', 'NJ', 'ME', 'MA', 'GA', 'ND']
 
Poor Network connectivity states in both area 415, 408 :- ['OR', 'NV', 'MA', 'GA', 'NC', 'KS']


In [74]:
# Assigning Customers data with no international plan and having voice mail plan state list in area 510
df_true_voice_510_states=list(df_true_voice_510['State'].unique())
print("States list Customers data with no international plan and having voice mail plan state list in area 510 :-",df_true_voice_510_states)
print(" ")
# Assigning customers data  with both international plan and voice mail plan state list in area 510
df_true_yes_510_states=list(df_true_yes_510['State'].unique())
print("States list customers data  with both international plan and voice mail plan state list in area 510 :-",df_true_yes_510_states)
print(" ")
# Getting Common  Network Maintainance states based on our hypothetical logic from area 510
maintainence_states=set(df_true_voice_510_states).intersection(set(df_true_yes_510_states))
print("Network needs to be maintenanced in states :-",list(maintainence_states))

States list Customers data with no international plan and having voice mail plan state list in area 510 :- ['MS', 'FL', 'WV', 'UT', 'NJ', 'TN', 'MN', 'SC', 'MT']
 
States list customers data  with both international plan and voice mail plan state list in area 510 :- ['ME', 'SD', 'VT', 'NV', 'OK', 'NJ', 'TX', 'WV', 'ID', 'MI']
 
Network needs to be maintenanced in states :- ['NJ', 'WV']


In [75]:
print(f"Poor Netwok States are {list(poor_network_states)} and Network needs to be maintenanced in states {list(maintainence_states)}.")

Poor Netwok States are ['OR', 'NV', 'MA', 'GA', 'NC', 'KS'] and Network needs to be maintenanced in states ['NJ', 'WV'].
